### PR TITLE
Pr/matching

### PR DESCRIPTION
**Pull request by zfred219**

### CODE REVIEW

**Comment by EnjiaXu** [comment]
Most files here are already in the previous PRs (need to include them to pass CI).
We will comment on which files to ignore.

File: app/assets/stylesheets/matching.scss [hunk 1]
```diff
@@ -0,0 +1,182 @@
...
+    width: 100%;
+  }
+}
+
+.card {
+  min-height: 120px;
+}
+
+  /* ~~~~~~~ INIT. BTN ~~~~~~~ */
+
```
Comment (zfred219) [inline app/assets/stylesheets/matching.scss:104]: Comment out another style of matching page

File: features/user_view.feature [hunk 2]
```diff
@@ -32,6 +32,7 @@ Scenario: User view displays the correct names
     Given I am on the Users page
     Then I should see "Joe Deatrick"
     And I should see "Armando Fox"
+		And I should see "GitHub Username"
```
Comment (zfred219) [inline features/user_view.feature:35]: Though it seems it's not aligned here, but the format it's correct locally

File: features/user_coach_auth.feature [hunk 3]
```diff
@@ -177,15 +156,8 @@ Scenario: Sad Path - Cannot create new iteration, edit and destroy existing iter
...
-  And I follow "app1"
-  And I follow "2017-03-25"
-  And I follow "Add Iteration"
-  Then I should see "Error: Only Staff can create, edit and destroy apps, engagements and iterations"
-  And I press "Edit" for "2017-04-14"
-  Then I should see "Error: Only Staff can create, edit and destroy apps, engagements and iterations"
-  And I press "Destroy" for "2017-04-14"
-  Then I should see "Error: Only Staff can create, edit and destroy apps, engagements and iterations"
+  And I am on the Apps page
+  Then I should see "You do not have access to that page."
```
Comment (zfred219) [inline features/user_coach_auth.feature:160]: Already included in previous PR, but to pass CI, we have to include modified tests.

Please ignore this file.

File: features/type_users.feature [hunk 4]
```diff
@@ -93,7 +93,7 @@ Scenario: User can submit a create form that includes user type and SID for Stud
     | User Name         | Fakeuser              |
     | Email             | fakeuser@berkeley.edu |
     | Preferred Contact | 555-555-5555          |
-    | Github Uid        | fakegithubuid         |
+    | GitHub Username   | fakegithubuid         |
```
Comment (zfred219) [inline features/type_users.feature:96]: Already included in previous PR, but to pass CI, we have to include modified tests.

Please ignore this file.

File: features/support/env.rb [hunk 5]
```diff
@@ -52,9 +52,14 @@
 #     DatabaseCleaner.strategy = :transaction
 #   end
 #
-
+options = {
+  :window_size => [6000, 3000]
+}
```
Comment (zfred219) [inline features/support/env.rb:57]: Already included in previous PR, but to pass CI, we have to resize window for webdriver

Please ignore this file.

File: features/step_definitions/search_steps.rb [hunk 6]
```diff
@@ -19,6 +19,14 @@
...
 end
 
+Given (/^the following engagements exist for "([^"]*)":$/) do |app_name, table|
+  @app = App.find_by_name(app_name)
+  table.hashes.each do |hash|
+		@app.engagements.create(hash)
+	end
+  @app.save!
+end
+
```
Comment (zfred219) [inline features/step_definitions/search_steps.rb:29]: Already included in previous PR, but to pass CI, we have to include new steps.

Please ignore this file.

File: features/step_definitions/matching_steps.rb [hunk 7]
```diff
@@ -0,0 +1,62 @@
+
```
Comment (zfred219) [inline features/step_definitions/matching_steps.rb:1]: All the step definitions for matching

File: features/step_definitions/clear_form_steps.rb [hunk 8]
```diff
@@ -20,6 +20,7 @@
 When /^(?:|I )fill in "([^"]*)" with "([^"]*)" using ckeEditor rich editor$/ do |field, value|
```
Comment (zfred219) [inline features/step_definitions/clear_form_steps.rb:20]: Please ignore this file.

File: features/step_definitions/auth_user_steps.rb [hunk 9]
```diff
@@ -1,3 +1,30 @@
-When /^(?:|I )am logged in as "([^"]*)"$/ do |user|
-  steps %Q{When PENDING:}
-end
\ No newline at end of file
+
```
Comment (zfred219) [inline features/step_definitions/auth_user_steps.rb:1]: Step definitions for different type of users

File: features/search.feature [hunk 10]
```diff
@@ -6,10 +6,10 @@ Feature: searching to quickly find apps according to name/organization/descripti
```
Comment (zfred219) [inline features/search.feature:6]: Please ignore this file.

File: features/public_search.feature [hunk 11]
```diff
@@ -24,13 +24,13 @@ Background: Logged Out
 
     And I'm am on the root page
 
-Scenario: No keyword
-    Given I publicly search for ""
-    Then I should not see "app 1"
+#Scenario: No keyword
```
Comment (zfred219) [inline features/public_search.feature:27]: Already included in previous PR, but to pass CI, we have to include modified tests.

Please ignore this file.

File: features/pagination.feature [hunk 12]
```diff
@@ -52,17 +52,17 @@ Background: Logged in
```
Comment (zfred219) [inline features/pagination.feature:52]: Already included in previous PR, but to pass CI, we have to include modified tests.

Please ignore this file.

File: app/assets/stylesheets/matching.scss [hunk 13]
```diff
@@ -0,0 +1,182 @@
+// Place all the styles related to the matching controller here.
```
Comment (zfred219) [inline app/assets/stylesheets/matching.scss:1]: **Matching**
- Matching related CSS

File: app/assets/stylesheets/progress.scss [hunk 14]
```diff
@@ -0,0 +1,95 @@
+.progress{
```
Comment (zfred219) [inline app/assets/stylesheets/progress.scss:1]: **Matching**
- Matching related CSS for progress page

File: app/controllers/matching_controller.rb [hunk 15]
```diff
@@ -0,0 +1,323 @@
+class MatchingController < ApplicationController
```
Comment (zfred219) [inline app/controllers/matching_controller.rb:1]: **Matching**
- Matching related controller and method

File: app/models/engagement.rb [hunk 16]
```diff
@@ -1,6 +1,7 @@
 class Engagement < ActiveRecord::Base
```
Comment (zfred219) [inline app/models/engagement.rb:1]: **Matching**
- Matching Model

File: app/models/engagement.rb [hunk 17]
```diff
@@ -1,6 +1,7 @@
...
   belongs_to :app
   belongs_to :coach, class_name: 'User'
+  belongs_to :matching
```
Comment (zfred219) [inline app/models/engagement.rb:4]: **Matching**
- Matching, Engagement relationship

File: app/models/matching.rb [hunk 18]
```diff
@@ -0,0 +1,273 @@
+class Matching < ActiveRecord::Base
```
Comment (zfred219) [inline app/models/matching.rb:1]: **Matching**
- Matching model

File: app/views/matching/index.html.haml [hunk 19]
```diff
@@ -0,0 +1,73 @@
+.page-header.jumbotron.mt-4.my-4
```
Comment (zfred219) [inline app/views/matching/index.html.haml:1]: **Matching**
- Matching index/main page

File: app/views/matching/new.html.haml [hunk 20]
```diff
@@ -0,0 +1,41 @@
+%h1.jumbotron
```
Comment (zfred219) [inline app/views/matching/new.html.haml:1]: **Matching**
- Create new matching view

File: app/views/matching/progress.html.haml [hunk 21]
```diff
@@ -0,0 +1,130 @@
+!!!
```
Comment (zfred219) [inline app/views/matching/progress.html.haml:1]: **Matching**
- Matching progress view
- Related javascripts

File: app/views/matching/result.html.haml [hunk 22]
```diff
@@ -0,0 +1,123 @@
+!!!
```
Comment (zfred219) [inline app/views/matching/result.html.haml:1]: **Matching**
- Matching ranking result view
- Related javascript

File: app/views/matching/show.html.haml [hunk 23]
```diff
@@ -0,0 +1,139 @@
+!!!
```
Comment (zfred219) [inline app/views/matching/show.html.haml:1]: **Matching**
- Matching ranking/show view
- Related javascript

File: config/initializers/assets.rb [hunk 24]
```diff
@@ -9,3 +9,8 @@
 # Precompile additional assets.
 # application.js, application.css, and all non-JS/CSS in app/assets folder are already added.
 # Rails.application.config.assets.precompile += %w( search.js )
+# Rails.application.config.assets.precompile += %w( main_page.js )
+Rails.application.config.assets.precompile += %w( apps.scss )
+Rails.application.config.assets.precompile += %w( matching.css )
```
Comment (zfred219) [inline config/initializers/assets.rb:14]: **Matching**
- Precompile matching scss

File: config/routes.rb [hunk 25]
```diff
@@ -1,4 +1,21 @@
...
+  get '/matching/:matching_id/progress' => 'matching#progress', :as => 'matching_progress'
+  post '/matching/:matching_id/engagement/:engagement_id/store' => 'matching#store'
+  get '/matching/:matching_id/result' => 'matching#result', :as => 'matching_result'
+  delete '/matching/:matching_id' => 'matching#destroy', :as => 'delete_matching'
+  post '/matching/:matching_id/result/final_edit' => 'matching#final_edit'
+  get '/matching/:matching_id/result/finalize' => 'matching#finalize', :as => 'matching_finalize'
+  post '/matching/:matching_id/engagement/create' => 'matching#create_engagement'
+  post '/matching/:matching_id/engagement/:engagement_id/update' => 'matching#update_engagement'
+  delete '/matching/:matching_id/engagement/:engagement_id' => 'matching#delete_engagement', :as => 'delete_engagement'
+  post '/matching/:matching_id/apps/update' => 'matching#update_apps'
```
Comment (zfred219) [inline config/routes.rb:17]: **Matching**
- Matching related routes

File: db/migrate/20210328005618_create_matchings.rb [hunk 26]
```diff
@@ -0,0 +1,15 @@
...
+      t.text :projects
+      t.text :preferences
+      t.text :description
+      t.text :result
+      t.string :status, null: false, default: 'Collecting responses'
+      t.text :last_edit_users
+      t.timestamps :last_edit_time
+    end
+  end
+end
```
Comment (zfred219) [inline db/migrate/20210328005618_create_matchings.rb:15]: **Matching**
- Matching table migration file

File: db/schema.rb [hunk 27]
```diff
@@ -93,6 +95,18 @@
...
+    t.datetime "updated_at"
+    t.string   "name"
+    t.text     "projects"
+    t.text     "preferences"
+    t.text     "description"
+    t.text     "result"
+    t.string   "status",          default: "Collecting responses", null: false
+    t.text     "last_edit_users"
+  end
+
```
Comment (zfred219) [inline db/schema.rb:109]: **Matching**
- Matching schema

File: db/schema.rb [hunk 28]
```diff
@@ -76,6 +76,8 @@
     t.string   "final_rating"
     t.string   "final_comments"
     t.string   "features"
+    t.string   "semester"
+    t.integer  "matching_id"
```
Comment (zfred219) [inline db/schema.rb:80]: **Matching**
- Engagement needs matching_id to identify matching

File: app/controllers/application_controller.rb [hunk 29]
```diff
@@ -97,4 +98,41 @@ def get_reviewed_apps
...
+      begin
+        redirect_to :back, alert: "You do not have access to that page."
+      rescue Exception
+        redirect_to "/", alert: "You do not have access to that page."
+      end
+    end
+  end
+
+  # redirects users to corresponding matching page
+  def auth_matching
```
Comment (zfred219) [inline app/controllers/application_controller.rb:115]: ** Matching **
- Check if current user is authorized or not for accessing

File: app/assets/stylesheets/application.scss [hunk 30]
```diff
@@ -7,4 +7,10 @@
  */
```
Comment (zfred219) [inline app/assets/stylesheets/application.scss:7]: Ignore this file (Included in a previous PR).

File: app/assets/stylesheets/users.scss [hunk 31]
```diff
@@ -0,0 +1,67 @@
+
```
Comment (zfred219) [inline app/assets/stylesheets/users.scss:1]: Ignore this file (Included in a previous PR).

File: app/controllers/apps_controller.rb [hunk 32]
```diff
@@ -1,6 +1,5 @@
 class AppsController < ApplicationController
```
Comment (zfred219) [inline app/controllers/apps_controller.rb:1]: Ignore this file (Included in a previous PR).

File: app/controllers/my_projects_controller.rb [hunk 33]
```diff
@@ -8,9 +8,69 @@ def index
         orgs = Org.for_user(user.id)
```
Comment (zfred219) [inline app/controllers/my_projects_controller.rb:8]: Ignore this file (Included in a previous PR).

File: app/controllers/orgs_controller.rb [hunk 34]
```diff
@@ -1,15 +1,16 @@
 class OrgsController < ApplicationController
```
Comment (zfred219) [inline app/controllers/orgs_controller.rb:1]: Ignore this file (Included in a previous PR).

File: app/controllers/search_controller.rb [hunk 35]
```diff
@@ -3,7 +3,7 @@ class SearchController < ApplicationController
```
Comment (zfred219) [inline app/controllers/search_controller.rb:3]: Ignore this file (Included in a previous PR).

File: app/controllers/session_controller.rb [hunk 36]
```diff
@@ -1,5 +1,6 @@
 class SessionController < ApplicationController
```
Comment (zfred219) [inline app/controllers/session_controller.rb:1]: Ignore this file (Included in a previous PR).

File: app/controllers/users_controller.rb [hunk 37]
```diff
@@ -1,11 +1,14 @@
 class UsersController < ApplicationController
```
Comment (zfred219) [inline app/controllers/users_controller.rb:1]: Ignore this file (Included in a previous PR).

File: app/models/user.rb [hunk 38]
```diff
@@ -1,3 +1,4 @@
+require 'csv'
```
Comment (zfred219) [inline app/models/user.rb:1]: Ignore this file (Included in a previous PR).

File: app/views/my_projects/index.html.haml [hunk 39]
```diff
@@ -1,35 +1,56 @@
-.page-header
-  %h1
-    My Projects
-  .help-block Click an App name to see more information
+!!!
```
Comment (zfred219) [inline app/views/my_projects/index.html.haml:1]: Ignore this file (Included in a previous PR).

File: app/views/users/_form.html.haml [hunk 40]
```diff
@@ -4,7 +4,7 @@
   = f.text_field :name, label: 'User Name'
```
Comment (zfred219) [inline app/views/users/_form.html.haml:4]: Ignore this file (Included in a previous PR).

File: app/views/users/index.html.haml [hunk 41]
```diff
@@ -1,7 +1,12 @@
+%csrf_meta_tag
```
Comment (zfred219) [inline app/views/users/index.html.haml:1]: Ignore this file (Included in a previous PR).

File: app/views/users/show.html.haml [hunk 42]
```diff
@@ -17,7 +17,7 @@
             %strong Email:
```
Comment (zfred219) [inline app/views/users/show.html.haml:17]: Ignore this file (Included in a previous PR).

File: db/migrate/20210303105028_add_semester_to_iterations.rb [hunk 43]
```diff
@@ -0,0 +1,5 @@
+class AddSemesterToIterations < ActiveRecord::Migration
```
Comment (zfred219) [inline db/migrate/20210303105028_add_semester_to_iterations.rb:1]: Ignore this file (Included in a previous PR).

File: db/migrate/20210414075312_add_matching_id_to_engagements.rb [hunk 44]
```diff
@@ -0,0 +1,5 @@
+class AddMatchingIdToEngagements < ActiveRecord::Migration
+  def change
+    add_column :engagements, :matching_id, :integer
+  end
+end
```
Comment (zfred219) [inline db/migrate/20210414075312_add_matching_id_to_engagements.rb:5]: **Matching**
- Add matching_id to engagement migration

File: db/orgs.csv [hunk 45]
```diff
@@ -0,0 +1,3 @@
+name,description,url,contact_id,address_line_1,address_line_2,city_state_zip,phone
```
Comment (zfred219) [inline db/orgs.csv:1]: Ignore this file (Included in a previous PR).

File: db/user_orgs.csv [hunk 46]
```diff
@@ -103,4 +103,15 @@ lar@berkeley.edu,Professor Larry Rosenthal,student,lar,UC Berkeley Goldman Schoo
 laura.montagna@gmail.com,Laura Montagna,student,laura.montagna,Youth Zone Del Norte,,
```
Comment (zfred219) [inline db/user_orgs.csv:103]: Ignore this file (Included in a previous PR).

File: features/app_features.feature [hunk 47]
```diff
@@ -50,7 +50,7 @@ Scenario: A user can add initial features while creating apps
     When I fill in "App Description" with "Fake app description"
```
Comment (zfred219) [inline features/app_features.feature:50]: Ignore this file (Included in a previous PR).

File: features/app_json.feature [hunk 48]
```diff
@@ -15,36 +15,37 @@ Background: apps with different status exist
         | app3  | test3       	| 1      | test3@test.com | in_use 		|
```
Comment (zfred219) [inline features/app_json.feature:15]: Ignore this file (Included in a previous PR).

File: features/apps_controller.feature [hunk 49]
```diff
@@ -28,10 +28,10 @@ Background: users, orgs and apps have been added to database
 Scenario: Login to github to create a new app
```
Comment (zfred219) [inline features/apps_controller.feature:28]: Ignore this file (Included in a previous PR).

File: features/create.feature [hunk 50]
```diff
@@ -57,7 +57,7 @@ Scenario: User can submit successfully if form is complete
         | User Name           | Fakeuser              |
```
Comment (zfred219) [inline features/create.feature:57]: Ignore this file (Included in a previous PR).

File: features/create_user.feature [hunk 51]
```diff
@@ -31,7 +31,7 @@ Scenario: There is a form on users page that I can use to create a user
     And I should see "User Name"
     And I should see "Email"
     And I should see "Preferred Contact"
-    And I should see "Github Uid"
+    And I should see "GitHub Username"
```
Comment (zfred219) [inline features/create_user.feature:34]: Ignore this file (Included in a previous PR).

File: features/engagement_sem.feature [hunk 52]
```diff
@@ -0,0 +1,47 @@
+Feature: Adding "semester" field for start date record
```
Comment (zfred219) [inline features/engagement_sem.feature:1]: Ignore this file (Included in a previous PR).

File: features/login_github_redirect.feature [hunk 53]
```diff
@@ -70,17 +70,17 @@ Scenario: login with Github from create page, after visiting orgs and users
 Scenario: Login to github to create a new app
```
Comment (zfred219) [inline features/login_github_redirect.feature:70]: Ignore this file (Included in a previous PR).

File: features/user_coach_auth.feature [hunk 54]
```diff
@@ -33,7 +33,7 @@ Scenario: Happy Path - Can create new app, edit and destroy existing app if I am
   And I press "Edit" for "app1"
```
Comment (zfred219) [inline features/user_coach_auth.feature:33]: Ignore this file (Included in a previous PR).

File: features/vetting_search_dropdown.feature [hunk 55]
```diff
@@ -5,6 +5,7 @@ I want to do a string matching within the dropdown menu
 
 Background: user and orgs have been added to database
```
Comment (zfred219) [inline features/vetting_search_dropdown.feature:6]: Ignore this file (Included in a previous PR).

File: spec/controllers/users_controller_spec.rb [hunk 56]
```diff
@@ -3,6 +3,7 @@
 describe UsersController, type: :controller do
```
Comment (zfred219) [inline spec/controllers/users_controller_spec.rb:3]: Ignore this file (Included in a previous PR).

File: features/matching_info.feature [hunk 57]
```diff
@@ -0,0 +1,102 @@
+Feature: Test Team Info page feature
```
Comment (zfred219) [inline features/matching_info.feature:1]: **Matching**
- Test regarding editing and info in progress page

File: features/matching_new.feature [hunk 58]
```diff
@@ -0,0 +1,96 @@
+Feature: Test Creating New Matching
```
Comment (zfred219) [inline features/matching_new.feature:1]: **Matching**
- Tests regarding creating new matching

File: spec/models/engagement_spec.rb [hunk 59]
```diff
@@ -1,4 +1,5 @@
 require 'rails_helper'
```
Comment (zfred219) [inline spec/models/engagement_spec.rb:1]: Ignore this file (Included in a previous PR).

File: spec/controllers/iterations_controller_spec.rb [hunk 60]
```diff
@@ -3,6 +3,7 @@
 describe IterationsController, type: :controller do
```
Comment (zfred219) [inline spec/controllers/iterations_controller_spec.rb:3]: Ignore this file (Included in a previous PR).

File: features/step_definitions/user_auth_steps.rb [hunk 61]
```diff
@@ -0,0 +1,10 @@
+Given /^I try to logged in with my github account$/ do
```
Comment (zfred219) [inline features/step_definitions/user_auth_steps.rb:1]: Ignore this file (Included in a previous PR).

File: features/step_definitions/login_steps.rb [hunk 62]
```diff
@@ -3,6 +3,6 @@
 end
```
Comment (zfred219) [inline features/step_definitions/login_steps.rb:3]: Ignore this file (Included in a previous PR).

File: features/matching_ranking.feature [hunk 63]
```diff
@@ -0,0 +1,49 @@
+Feature: Test Ranking page feature
```
Comment (zfred219) [inline features/matching_ranking.feature:1]: **Matching**
- Tests regarding ranking page

**Pull request merged**